File: vibebot/workers/delivery_worker.py
import json
import logging
import time

from vibebot.core.db import (
    get_deliverable_messages,
    mark_message_failed,
    mark_message_retry,
    mark_message_sent,
)
from vibebot.core.slack_sender import SlackSender

log = logging.getLogger(__name__)


def run(conn) -> None:
    """Stage 3: Deliver all pending outbound messages with retry and failure notification."""
    messages = get_deliverable_messages(conn)
    if not messages:
        log.info("No pending messages to deliver.")
        return
    for msg in messages:
        _deliver_with_retry(conn, msg)


def _deliver_with_retry(conn, msg) -> None:
    max_retries = msg["max_retries"]
    last_error = ""

    for attempt in range(msg["retry_count"], max_retries):
        try:
            SlackSender().send(json.loads(msg["payload"]))
            mark_message_sent(conn, msg["id"])
            log.info(
                "Delivered message id=%d (%s) on attempt %d.",
                msg["id"], msg["message_type"], attempt + 1,
            )
            return
        except Exception as e:
            last_error = str(e)
            mark_message_retry(conn, msg["id"], last_error)
            wait = 2 ** attempt  # exponential backoff: 1s, 2s, 4s ...
            log.warning(
                "Delivery attempt %d/%d failed for message id=%d: %s. Retrying in %ds.",
                attempt + 1, max_retries, msg["id"], last_error, wait,
            )
            time.sleep(wait)

    # All attempts exhausted
    mark_message_failed(conn, msg["id"], last_error)
    log.error(
        "Message id=%d (%s) permanently failed after %d attempt(s).",
        msg["id"], msg["message_type"], max_retries,
    )
    _notify_failure(msg["message_type"], max_retries, last_error)


def _notify_failure(message_type: str, attempts: int, error: str) -> None:
    blocks = [
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": (
                    f"⚠️ *VibeBot delivery failed*\n"
                    f"*Type:* {message_type}\n"
                    f"*After {attempts} attempt(s)*\n"
                    f"*Error:* `{error}`"
                ),
            },
        }
    ]
    try:
        SlackSender().send(blocks)
    except Exception:
        log.error("Could not send failure notification to Slack.")

File: src/workers/deliveryWorker.ts
import Database from 'better-sqlite3';
import {
  getDeliverableMessages,
  markMessageFailed,
  markMessageRetry,
  markMessageSent,
  OutboundMessage,
} from '../core/db';
import { SlackSender } from '../core/slackSender';

export async function run(db: Database.Database): Promise<void> {
  const messages = getDeliverableMessages(db);
  if (!messages.length) {
    console.info('No pending messages to deliver.');
    return;
  }
  for (const msg of messages) {
    await deliverWithRetry(db, msg);
  }
}

async function deliverWithRetry(db: Database.Database, msg: OutboundMessage): Promise<void> {
  const maxRetries = msg.max_retries;
  let lastError = '';

  for (let attempt = msg.retry_count; attempt < maxRetries; attempt++) {
    try {
      await new SlackSender().send(JSON.parse(msg.payload));
      markMessageSent(db, msg.id);
      console.info(`Delivered message id=${msg.id} (${msg.message_type}) on attempt ${attempt + 1}.`);
      return;
    } catch (e) {
      lastError = String(e instanceof Error ? e.message : e);
      markMessageRetry(db, msg.id, lastError);
      const wait = Math.pow(2, attempt) * 1000;
      console.warn(
        `Delivery attempt ${attempt + 1}/${maxRetries} failed for message id=${msg.id}: ${lastError}. Retrying in ${wait / 1000}s.`,
      );
      await new Promise((resolve) => setTimeout(resolve, wait));
    }
  }

  markMessageFailed(db, msg.id, lastError);
  console.error(
    `Message id=${msg.id} (${msg.message_type}) permanently failed after ${maxRetries} attempt(s).`,
  );
  await notifyFailure(msg.message_type, maxRetries, lastError);
}

async function notifyFailure(messageType: string, attempts: number, error: string): Promise<void> {
  const blocks = [
    {
      type: 'section',
      text: {
        type: 'mrkdwn',
        text: `⚠️ *VibeBot delivery failed*\n*Type:* ${messageType}\n*After ${attempts} attempt(s)*\n*Error:* \`${error}\``,
      },
    },
  ];
  try {
    await new SlackSender().send(blocks);
  } catch {
    console.error('Could not send failure notification to Slack.');
  }
}
